src/jit/lower.cpp

```diff
@@ -3953,6 +4025,16 @@ void Lowering::DoPhase()
         currBlock = block;
         comp->compCurBB = block;
 
+#if !defined(_TARGET_64BIT_)
+        // Decompose TYP_LONG nodes
+        for (stmt = block->bbTreeList; stmt != nullptr; stmt = stmt->gtNext)
+        {
+            comp->compCurStmt = stmt;
+            comp->fgWalkTreePost(&stmt->gtStmt.gtStmtExpr, &Lowering::DecompNodeHelper, this, true);
+            stmt = comp->compCurStmt;
+        }
+#endif
+
         /* Walk the statement trees in this basic block */
```

Comment:
Actually, I think a better solution is to move the addressing mode realization earlier in the phase order. We already have CSE determining whether an addressing mode is likely to be created, before it decides whether to create a CSE. This is something we've discussed, but it hasn't bubbled up to the top of the list.